crates/bls-crypto/src/bls/signature.rs

```diff
@@ -66,6 +66,24 @@ impl Signature {
             .into()
     }
 
+    /// Computes the dot product of a vector of signatures and a vector of exponents.
+    pub fn batch(r: &[Fr], signatures: &[&Signature]) -> Signature {
+        // Get everything in the right types
+        let projective_elements = signatures
+            .iter()
+            .map(|s| s.as_ref().clone())
+            .collect::<Vec<G1Projective>>();
+
+        let bases = G1Projective::batch_normalization_into_affine(&projective_elements);
+
+        let bigint_scalars: Vec<BigInteger256> = r
```

Comment:
Nit: in many places, you can replace `iter` and its variants with `cfg_iter!` and its variants from arkworks, for configurable parallelization. Not super important here.